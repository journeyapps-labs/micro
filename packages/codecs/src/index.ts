export * from './parsers';
export * from './codecs';
