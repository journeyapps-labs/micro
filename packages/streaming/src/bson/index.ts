export * from "./buffer-array";
export * from "./constants";
export * from "./encoder";
export * from "./decoder";
export * from "./header";
