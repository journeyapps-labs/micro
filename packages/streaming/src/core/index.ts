export * from './backpressure';
export * from './transformers';
export * from './node-utils';
export * from './utils';
