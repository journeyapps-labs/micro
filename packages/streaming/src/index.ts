export * as compat from './core/cross-stream';
export * as middleware from './middleware';
export * as bson from './bson';
export * from './core';
