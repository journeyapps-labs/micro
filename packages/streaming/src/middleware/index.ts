export * from './streamed-body-parser-v1';
export * from './streamed-body-parser-v2';
