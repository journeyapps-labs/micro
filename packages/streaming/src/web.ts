export * as compat from "./core/cross-stream";
export * as bson from "./bson";

export * from "./core/transformers";
export * from "./core/utils";
