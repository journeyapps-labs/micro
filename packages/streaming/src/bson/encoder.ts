import * as stream from "../core/cross-stream";
import * as constants from "./constants";
import * as bson from "bson";

export type BSONStreamEncoderParams<T> = {
  serialize_options?: bson.SerializeOptions;
  sendTerminatorOnEnd?: boolean;
  writableStrategy?: QueuingStrategy<T | undefined>;
  readableStrategy?: QueuingStrategy<Buffer | undefined>;
};

export const createBSONStreamEncoder = <T extends {} = any>(
  params?: BSONStreamEncoderParams<T>,
) => {
  let readableStrategy = params?.readableStrategy;
  if (!readableStrategy) {
    readableStrategy = new stream.ByteLengthStrategy({
      highWaterMark: 1024 * 16,
    });
  }

  return new stream.Transform<T, Buffer>(
    {
      transform(chunk, controller) {
        controller.enqueue(
          Buffer.from(bson.serialize(chunk, params?.serialize_options)),
        );
      },
      flush(controller) {
        if (params?.sendTerminatorOnEnd ?? true) {
          controller.enqueue(constants.TERMINATOR);
        }
      },
    },
    params?.writableStrategy,
    readableStrategy,
  );
};
