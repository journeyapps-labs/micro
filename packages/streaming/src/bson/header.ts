import * as buffer_array from './buffer-array';
import * as bson from 'bson';

export type DecodedResponse<T> = {
  header: T;
  stream: AsyncIterable<Buffer>;
};

export type ExtractHeaderParams = {
  deserialize_options?: bson.DeserializeOptions;
};
export const extractHeaderFromStream = async <T = any>(
  input_stream: AsyncIterable<Buffer>,
  params?: ExtractHeaderParams
): Promise<DecodedResponse<T>> => {
  const iterator = input_stream[Symbol.asyncIterator]();

  const buffer = buffer_array.createReadableBufferArray();
  let frame_size: number | null = null;

  async function* resplice(data: Buffer | null, iterator: AsyncIterator<Buffer>) {
    if (data) {
      yield data;
    }

    while (true) {
      const next = await iterator.next();
      if (next.done) {
        return;
      }

      yield next.value;
    }
  }

  while (true) {
    const chunk = await iterator.next();
    if (chunk.done) {
      throw new Error('Stream did not complete successfully');
    }

    buffer.push(chunk.value);

    if (frame_size === null) {
      frame_size = buffer.peek(4)?.readInt32LE(0) || null;
    }
    if (frame_size === null) {
      continue;
    }

    const frame = buffer.read(frame_size);
    if (!frame) {
      continue;
    }

    const header = bson.deserialize(frame, {
      promoteBuffers: true,
      ...(params?.deserialize_options || {})
    });

    return {
      header: header as T,
      stream: resplice(buffer.read(buffer.size()), iterator)
    };
  }
};

export async function* prependHeaderToStream<T>(
  header: any,
  input_stream: Iterable<T> | AsyncIterable<T>
): AsyncGenerator<T | Uint8Array> {
  yield bson.serialize(header);
  yield* input_stream;
}
