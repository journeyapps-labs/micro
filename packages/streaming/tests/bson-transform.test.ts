import { describe, test, expect } from 'vitest';

import * as micro_streaming from '../src';
import * as stream from 'stream/web';
import * as bson from 'bson';

describe('bson-transformer', () => {
  test('it successfully reads data from chunks array', () => {
    const chunks = Array.from(Array(10).keys()).map((_, i) => Buffer.alloc(i));

    expect(micro_streaming.bson.readBufferFromChunks(chunks, 50)).toBe(null);
    expect(micro_streaming.bson.readBufferFromChunks(chunks, 5)?.chunks_read).toBe(4);
    expect(micro_streaming.bson.readBufferFromChunks(chunks, 2)?.buffer.length).toBe(3);
  });

  test('it successfully reads data from chunks array, modifying original', () => {
    const chunks = Array.from(Array(10).keys()).map((_, i) => Buffer.alloc(i));

    const read1 = micro_streaming.bson.readBufferFromChunksAndModify(chunks, 1);
    expect(read1?.length).toBe(1);
    expect(chunks.length).toBe(8);

    const read2 = micro_streaming.bson.readBufferFromChunksAndModify(chunks, 1);
    expect(read2?.length).toBe(1);
    expect(chunks.length).toBe(8);
    expect(chunks[0].length).toBe(1);
  });

  test('it successfully deserializes streamed bson data', async () => {
    const source = new stream.ReadableStream({
      start(controller) {
        controller.enqueue(bson.serialize({ a: 'b' }));
        controller.enqueue(bson.serialize({ c: 'd' }));
        controller.enqueue(bson.serialize({ e: 'f' }));
        controller.enqueue(micro_streaming.bson.TERMINATOR);
        controller.close();
      }
    });

    const sink = source.pipeThrough(micro_streaming.bson.createBSONStreamDecoder());
    expect(await micro_streaming.drain(sink)).toMatchSnapshot();
  });

  test('it successfully serializes objects to bson stream', async () => {
    const source = new stream.ReadableStream({
      start(controller) {
        controller.enqueue({ a: 'b' });
        controller.enqueue({ c: 'd' });
        controller.enqueue({ e: 'f' });
        controller.close();
      }
    });

    const sink = source.pipeThrough(micro_streaming.bson.createBSONStreamEncoder());
    expect(Buffer.concat(await micro_streaming.drain(sink))).toMatchSnapshot();
  });

  test('end-to-end streaming', async () => {
    const output = micro_streaming
      .readableFrom([{ a: 'b' }, { c: 'd' }, { e: 'f' }])
      .pipeThrough(micro_streaming.bson.createBSONStreamEncoder())
      .pipeThrough(micro_streaming.bson.createBSONStreamDecoder());

    expect(await micro_streaming.drain(output)).toMatchSnapshot();
  });
});
