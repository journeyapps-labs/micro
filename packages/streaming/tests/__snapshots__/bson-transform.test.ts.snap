// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bson-transformer > end-to-end streaming 1`] = `
[
  {
    "a": "b",
  },
  {
    "c": "d",
  },
  {
    "e": "f",
  },
]
`;

exports[`bson-transformer > it successfully deserializes streamed bson data 1`] = `
[
  {
    "a": "b",
  },
  {
    "c": "d",
  },
  {
    "e": "f",
  },
]
`;

exports[`bson-transformer > it successfully serializes objects to bson stream 1`] = `
{
  "data": [
    14,
    0,
    0,
    0,
    2,
    97,
    0,
    2,
    0,
    0,
    0,
    98,
    0,
    0,
    14,
    0,
    0,
    0,
    2,
    99,
    0,
    2,
    0,
    0,
    0,
    100,
    0,
    0,
    14,
    0,
    0,
    0,
    2,
    101,
    0,
    2,
    0,
    0,
    0,
    102,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;
